src/gui/log/logmodel.cpp

```diff
@@ -0,0 +1,192 @@
+ * exception statement from your version.
+ */
+
+#include "logmodel.h"
+
+#include <QApplication>
+#include <QDateTime>
+#include <QColor>
+#include <QPalette>
+
+#include "base/global.h"
+
+namespace
+{
+    const int MAX_VISIBLE_MESSAGES = 20000;
+}
+
+BaseLogModel::Message::Message(const QString &time, const QString &message, const QColor &foreground, const Log::MsgType type)
+    : m_time(time)
+    , m_message(message)
+    , m_foregroundRole(foreground)
+    , m_userRole(type)
+{
+}
+
+QVariant BaseLogModel::Message::time() const
+{
+    return m_time;
+}
+
+QVariant BaseLogModel::Message::message() const
+{
+    return m_message;
+}
+
+QVariant BaseLogModel::Message::foregroundRole() const
+{
+    return m_foregroundRole;
+}
+
+QVariant BaseLogModel::Message::userRole() const
+{
+    return m_userRole;
+}
+
+BaseLogModel::BaseLogModel(QObject *parent)
+    : QAbstractListModel(parent)
+    , m_messages(MAX_VISIBLE_MESSAGES)
+{
+}
+
+int BaseLogModel::rowCount(const QModelIndex &) const
+{
+    return m_messages.size();
+}
+
+int BaseLogModel::columnCount(const QModelIndex &) const
+{
+    return 1;
+}
+
+QVariant BaseLogModel::data(const QModelIndex &index, const int role) const
+{
+    if (!index.isValid())
+        return {};
+
+    const int messageIndex = index.row();
+    if ((messageIndex < 0) || (messageIndex >= m_messages.size()))
+        return {};
+
+    const Message &message = m_messages[messageIndex];
+    switch (role) {
+    case Qt::ForegroundRole:
+        return message.foregroundRole();
+    case Qt::UserRole:
+        return message.userRole();
+    case TimeRole:
+        return message.time();
+    case MessageRole:
+        return message.message();
+    default:
+        return {};
+    }
+}
+
+void BaseLogModel::addNewMessage(const BaseLogModel::Message &message)
+{
+    // if row is inserted on filled up buffer, the size will not change but because of calling of beginInsertRows function
+    // we'll have ghost rows.
+    if (m_messages.size() == MAX_VISIBLE_MESSAGES) {
+        const int lastMessage = m_messages.size() - 1;
+        beginRemoveRows(QModelIndex(), lastMessage, lastMessage);
+        m_messages.pop_back();
+        endRemoveRows();
+    }
+
+    beginInsertRows(QModelIndex(), 0, 0);
+    m_messages.push_front(message);
+    endInsertRows();
+}
+
+void BaseLogModel::reset()
+{
+    beginResetModel();
+    m_messages.clear();
+    endResetModel();
+}
+
+LogMessageModel::LogMessageModel(QObject *parent)
+    : BaseLogModel(parent)
+{
+    for (const Log::Msg &msg : asConst(Logger::instance()->getMessages()))
+        handleNewMessage(msg);
+
+    // Logger::newLogMessage can be emitted from different threads, so to avoid need of synchronization we use Qt::QueuedConnection
+    connect(Logger::instance(), &Logger::newLogMessage, this, &LogMessageModel::handleNewMessage, Qt::QueuedConnection);
+}
+
+void LogMessageModel::handleNewMessage(const Log::Msg &log)
```

Comment:
"log" is incorrect name for this parameter. I would call it "message" and rename the next one to "messageText".